File: postgremq-ts/src/__tests__/metrics.test.ts
import { metrics as otelMetrics } from '@opentelemetry/api';
import { test, expect } from '@jest/globals';
import {
  InMemoryMetricExporter,
  MeterProvider,
  PeriodicExportingMetricReader,
  AggregationTemporality,
} from '@opentelemetry/sdk-metrics';
import { readFileSync } from 'fs';
import { join } from 'path';
import { Connection } from '../connection';
import { Message } from '../message';
import { createIsolatedTestConnection, getSharedTestDatabase } from './helpers';
import { LeaseLostError, QueueNotFoundError } from '../errors';

const contract = JSON.parse(
  readFileSync(join(__dirname, '../../../observability/client-contract.json'), 'utf8'),
);

function telemetry() {
  const exporter = new InMemoryMetricExporter(AggregationTemporality.CUMULATIVE);
  const reader = new PeriodicExportingMetricReader({ exporter, exportIntervalMillis: 60000 });
  return { exporter, provider: new MeterProvider({ readers: [reader] }) };
}

test('metrics contract with transactions, empty receive, redelivery and failed settlement', async () => {
  const iso = await createIsolatedTestConnection(await getSharedTestDatabase());
  const { exporter, provider } = telemetry();
  const c = new Connection({ pool: iso.pool, meterProvider: provider });
  try {
    await c.connect();
    await c.createTopic('metrics');
    await c.createQueue('q', 'metrics', false);
    await c.publish('metrics', {});
    const tx = await iso.pool.connect();
    try {
      await tx.query('BEGIN');
      await c.publishWithTransaction(tx, 'metrics', {});
      await tx.query('ROLLBACK');
      await expect(c.publish('missing', {})).rejects.toBeInstanceOf(QueueNotFoundError);
      const take = async () => {
        const rows = await c.consumeMessages('q', 30, 1);
        expect(rows).toHaveLength(1);
        const r = rows[0];
        return new Message(
          r.message_id,
          'q',
          r.payload,
          r.consumer_token,
          r.delivery_attempts,
          r.vt,
          r.published_at,
          () => {},
          {
            ack: c.ackMessage.bind(c),
            nack: c.nackMessage.bind(c),
            release: c.releaseMessage.bind(c),
            setVt: c.setMessageVt.bind(c),
          },
        );
      };
      let msg = await take();
      await msg.setVt(30);
      await tx.query('BEGIN');
      await msg.ackWithTransaction(tx);
      await tx.query('ROLLBACK');
      await expect(msg.ack()).rejects.toThrow('already been processed');
      await iso.pool.query(
        "UPDATE postgremq.queue_messages SET vt=clock_timestamp()-interval '1 second'",
      );
      msg = await take();
      await msg.nack();
      msg = await take();
      await msg.release();
      await iso.pool.query('DELETE FROM postgremq.queue_messages');
      expect(await c.consumeMessages('q', 30, 1)).toEqual([]);
      await iso.pool.query("SELECT postgremq.publish_message('metrics','{}')");
      msg = await take();
      await iso.pool.query(
        'UPDATE postgremq.queue_messages SET consumer_token=gen_random_uuid()::text',
      );
      await expect(msg.ack()).rejects.toBeInstanceOf(LeaseLostError);
    } finally {
      await tx.query('ROLLBACK');
      tx.release();
    }
    await provider.forceFlush();
    const scopes = exporter.getMetrics().flatMap((r) => r.scopeMetrics);
    expect(scopes).toHaveLength(1);
    expect(scopes[0].scope.name).toBe(contract.scope);
    expect(scopes[0].scope.version).toBe(contract.version);
    for (const metric of scopes[0].metrics)
      expect(metric.descriptor.unit).toBe(contract.metrics[metric.descriptor.name]);
    const operations = scopes[0].metrics.find(
      (m) => m.descriptor.name === 'messaging.client.operation.duration',
    )!;
    expect(operations.dataPoints).toHaveLength(contract.scenario_operations.length);
    for (const expected of contract.scenario_operations) {
      const attrs: Record<string, string | boolean> = {
        'messaging.system': 'postgremq',
        'messaging.operation.name': expected.name,
        'messaging.operation.type': expected.type,
        'messaging.destination.name': expected.destination,
        'postgremq.transaction': expected.transaction,
        ...(expected.error ? { 'error.type': expected.error } : {}),
      };
      const point = operations.dataPoints.find((p) =>
        Object.keys(attrs).every((k) => p.attributes[k] === attrs[k]),
      );
      expect(point).toBeDefined();
      expect(point!.attributes).toEqual(attrs);
      const histogram = point!.value as {
        count: number;
        sum: number;
        buckets: { boundaries: number[] };
      };
      expect(histogram.count).toBe(expected.count);
      expect(histogram.sum).toBeGreaterThanOrEqual(0);
      expect(histogram.buckets.boundaries).toEqual(contract.histogram_boundaries);
    }
    const sent = scopes[0].metrics.find(
      (m) => m.descriptor.name === 'messaging.client.sent.messages',
    )!;
    expect(sent.dataPoints).toHaveLength(contract.sent.length);
    for (const expected of contract.sent) {
      const attrs = {
        'messaging.system': 'postgremq',
        'messaging.operation.name': 'publish',
        'messaging.operation.type': 'send',
        'messaging.destination.name': expected.destination,
        'postgremq.transaction': expected.transaction,
        ...(expected.error ? { 'error.type': expected.error } : {}),
      };
      const point = sent.dataPoints.find(
        (p) =>
          p.attributes['messaging.destination.name'] === expected.destination &&
          p.attributes['postgremq.transaction'] === expected.transaction,
      );
      expect(point?.attributes).toEqual(attrs);
      expect(point?.value).toBe(expected.count);
    }
    const received = scopes[0].metrics.find(
      (m) => m.descriptor.name === 'messaging.client.consumed.messages',
    )!;
    expect(
      received.dataPoints.find((p) => p.attributes['postgremq.redelivered'] === false)!.value,
    ).toBe(contract.received.first);
    expect(
      received.dataPoints.find((p) => p.attributes['postgremq.redelivered'] === true)!.value,
    ).toBe(contract.received.redelivered);
  } finally {
    await c.close();
    await provider.shutdown();
    await iso.connection.close();
    await iso.dropDatabase();
  }
});

test.each(['', 'handler_error', 'cancelled'])(
  'handler metrics balance active callbacks: %s',
  async (outcome) => {
    const { HandlerConsumer } = await import('../handler-consumer');
    const { exporter, provider } = telemetry();
    const c = new Connection({ meterProvider: provider });
    let unblock!: () => void;
    let entered!: () => void;
    const gate = new Promise<void>((resolve) => {
      unblock = resolve;
    });
    const started = new Promise<void>((resolve) => {
      entered = resolve;
    });
    const msg = new Message(1, 'q', {}, 'token', 1, new Date(), new Date(), () => {}, {
      ack: async () => {},
      nack: async () => {},
      release: async () => {},
      setVt: async () => new Date(),
    });
    // Explicit settlement must not end processing metrics while the callback runs.
    await msg.ack();
    const hc = new HandlerConsumer(
      c,
      {} as import('../consumer').Consumer,
      async () => {
        entered();
        await gate;
        if (outcome === 'handler_error') throw new Error('private handler detail');
      },
      1,
    );
    const running = hc['runHandler'](msg);
    try {
      await started;
      await provider.forceFlush();
      const during = exporter
        .getMetrics()
        .flatMap((r) => r.scopeMetrics)
        .flatMap((s) => s.metrics);
      expect(
        during.find((m) => m.descriptor.name === 'postgremq.client.handlers.active')!.dataPoints[0]
          .value,
      ).toBe(1);
      expect(
        during.find((m) => m.descriptor.name === 'messaging.process.duration'),
      ).toBeUndefined();
      if (outcome === 'cancelled') msg._cancel();
      unblock();
      await running;
      exporter.reset();
      await provider.forceFlush();
      const after = exporter
        .getMetrics()
        .flatMap((r) => r.scopeMetrics)
        .flatMap((s) => s.metrics);
      expect(
        after.find((m) => m.descriptor.name === 'postgremq.client.handlers.active')!.dataPoints[0]
          .value,
      ).toBe(0);
      const process = after.find((m) => m.descriptor.name === 'messaging.process.duration')!;
      expect(process.descriptor.unit).toBe(contract.metrics[process.descriptor.name]);
      expect(process.dataPoints).toHaveLength(1);
      expect((process.dataPoints[0].value as { count: number }).count).toBe(1);
      expect(process.dataPoints[0].attributes['error.type'] ?? '').toBe(outcome);
    } finally {
      unblock();
      await running;
      await c.close();
      await provider.shutdown();
    }
  },
);

test('renewal metrics ignore deregistered deliveries and count retired entries once', async () => {
  const { exporter, provider } = telemetry();
  const c = new Connection({ meterProvider: provider });
  try {
    // Drive the existing renewal actor with a successful batch missing its row.
    // No database/poll timer is needed to test apply behavior.
    const e = {
      queue: 'q',
      id: 1,
      token: 't',
      vtSec: 30,
      threshold: 0.5,
      nextExtensionTime: new Date(0),
      expiresAt: new Date(Date.now() + 30000),
      cancel: () => {},
      onExtended: () => {},
    };
    c['extenderEntries'].set('q\0' + '1\0t', e);
    c.setVtBatchMulti = async () => [];
    await c['flushExtender']();
    await c['flushExtender']();
    await provider.forceFlush();
    const metric = exporter
      .getMetrics()
      .flatMap((r) => r.scopeMetrics)
      .flatMap((s) => s.metrics)
      .find((m) => m.descriptor.name === 'postgremq.client.renewal.lost')!;
    expect(metric.descriptor.unit).toBe(contract.metrics[metric.descriptor.name]);
    expect(metric.dataPoints[0].value).toBe(1);
    expect(metric.dataPoints[0].attributes).toEqual({
      'messaging.system': 'postgremq',
      'messaging.destination.name': 'q',
    });
  } finally {
    await c.close();
    await provider.shutdown();
  }
});

test('operation duration counts a retried publication once, including backoff', async () => {
  const iso = await createIsolatedTestConnection(await getSharedTestDatabase());
  const { exporter, provider } = telemetry();
  const c = new Connection({ pool: iso.pool, meterProvider: provider });
  try {
    await c.connect();
    await c.createTopic('metrics');
    await iso.pool.query(
      readFileSync(join(__dirname, '../../../observability/retry-once.sql'), 'utf8'),
    );
    await c.publish('metrics', {});
    expect(
      (await iso.pool.query('SELECT last_value FROM public.metrics_retry_probe')).rows[0]
        .last_value,
    ).toBe('2');
    await provider.forceFlush();
    const metrics = exporter
      .getMetrics()
      .flatMap((r) => r.scopeMetrics)
      .flatMap((s) => s.metrics);
    expect(metrics).toHaveLength(2);
    const duration = metrics.find(
      (m) => m.descriptor.name === 'messaging.client.operation.duration',
    )!;
    const sent = metrics.find((m) => m.descriptor.name === 'messaging.client.sent.messages')!;
    expect(sent.dataPoints).toHaveLength(2);
    expect(
      Object.fromEntries(sent.dataPoints.map((p) => [p.attributes['error.type'] ?? '', p.value])),
    ).toEqual({ '': 1, other: 1 });
    expect(duration.dataPoints).toHaveLength(1);
    const point = duration.dataPoints[0];
    expect(point.attributes['error.type']).toBeUndefined();
    expect((point.value as { count: number }).count).toBe(1);
    expect((point.value as { sum: number }).sum).toBeGreaterThanOrEqual(0.1);
  } finally {
    await c.close();
    await provider.shutdown();
    await iso.connection.close();
    await iso.dropDatabase();
  }
});

test.each([false, true])(
  'metrics are optional and provider remains application-owned: %s',
  async (enabled) => {
    const iso = await createIsolatedTestConnection(await getSharedTestDatabase());
    const { exporter, provider } = telemetry();
    expect(otelMetrics.setGlobalMeterProvider(provider)).toBe(true);
    const c = new Connection({ pool: iso.pool, ...(enabled ? { meterProvider: provider } : {}) });
    try {
      await c.connect();
      await c.createTopic('optional');
      await c.createQueue('q', 'optional', false);
      await c.publish('optional', {});
      const messages = await c.consumeMessages('q', 30, 1);
      expect(messages).toHaveLength(1);
      await c.ackMessage('q', messages[0].message_id, messages[0].consumer_token);
      // Locally rejected/unsent messages must not contribute to sent.messages.
      const circular: any = {};
      circular.self = circular;
      await expect(c.publish('optional', circular)).rejects.toThrow();
      await c.close();
      await expect(c.publish('optional', {})).rejects.toThrow();
      provider.getMeter('application').createCounter('application.probe').add(1);
      await provider.forceFlush();
      const scopes = exporter.getMetrics().flatMap((r) => r.scopeMetrics);
      expect(
        scopes.find((s) => s.scope.name === 'application')?.metrics[0].dataPoints[0].value,
      ).toBe(1);
      const client = scopes.find((s) => s.scope.name === 'postgremq');
      expect(!!client).toBe(enabled);
      if (enabled) {
        const sent = client!.metrics.find(
          (m) => m.descriptor.name === 'messaging.client.sent.messages',
        )!;
        expect(sent.dataPoints).toHaveLength(1);
        expect(sent.dataPoints[0].value).toBe(1);
      }
    } finally {
      otelMetrics.disable();
      await c.close();
      await provider.shutdown();
      await iso.connection.close();
      await iso.dropDatabase();
    }
  },
);

File: postgremq-go/metrics_test.go
package postgremq_go_test

import (
	"encoding/json"
	"os"
	"testing"

	postgremq "github.com/slavakl/postgremq/postgremq-go"
	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/metric/metricdata"
)

func TestMetricsContractWithTransactionsAndRedelivery(t *testing.T) {
	pool, ctx := setupTestConnection(t)
	reader := sdkmetric.NewManualReader()
	provider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(reader))
	defer func() { require.NoError(t, provider.Shutdown(ctx)) }()
	conn, err := postgremq.DialFromPool(pool, postgremq.WithMeterProvider(provider), postgremq.WithoutRetries())
	require.NoError(t, err)
	defer func() { require.NoError(t, conn.Close()) }()
	require.NoError(t, conn.CreateTopic(ctx, "metrics"))
	require.NoError(t, conn.CreateQueue(ctx, "q", "metrics", false))
	_, err = conn.Publish(ctx, "metrics", []byte(`{}`))
	require.NoError(t, err)
	tx, err := pool.Begin(ctx)
	require.NoError(t, err)
	_, err = conn.PublishWithTx(ctx, tx, "metrics", []byte(`{}`))
	require.NoError(t, err)
	require.NoError(t, tx.Rollback(ctx))
	_, err = conn.Publish(ctx, "missing", []byte(`{}`))
	require.ErrorIs(t, err, postgremq.ErrQueueNotFound)
	take := func() *postgremq.Message {
		msgs, e := conn.ConsumeMessages(ctx, "q", 1, 30)
		require.NoError(t, e)
		require.Len(t, msgs, 1)
		return msgs[0]
	}
	msg := take()
	_, err = msg.SetVT(ctx, 30)
	require.NoError(t, err)
	tx, err = pool.Begin(ctx)
	require.NoError(t, err)
	require.NoError(t, msg.AckWithTx(ctx, tx))
	require.NoError(t, tx.Rollback(ctx))
	// Local duplicate settlement does not execute an operation or emit one.
	require.ErrorIs(t, msg.Ack(ctx), postgremq.ErrLeaseLost)
	_, err = pool.Exec(ctx, "UPDATE postgremq.queue_messages SET vt=clock_timestamp()-interval '1 second'")
	require.NoError(t, err)
	msg = take()
	require.NoError(t, msg.Nack(ctx))
	msg = take()
	require.NoError(t, msg.Release(ctx))
	// No live messages left, but an empty receive is still a measured operation.
	_, err = pool.Exec(ctx, "DELETE FROM postgremq.queue_messages")
	require.NoError(t, err)
	msgs, err := conn.ConsumeMessages(ctx, "q", 1, 30)
	require.NoError(t, err)
	require.Empty(t, msgs)
	// A delivery whose server lease disappears produces a settlement error.
	_, err = pool.Exec(ctx, "SELECT postgremq.publish_message('metrics','{}')")
	require.NoError(t, err)
	lost := take()
	_, err = pool.Exec(ctx, "UPDATE postgremq.queue_messages SET consumer_token=gen_random_uuid()::text")
	require.NoError(t, err)
	require.ErrorIs(t, lost.Ack(ctx), postgremq.ErrLeaseLost)

	var data metricdata.ResourceMetrics
	require.NoError(t, reader.Collect(ctx, &data))
	raw, err := os.ReadFile("../observability/client-contract.json")
	require.NoError(t, err)
	var contract struct {
		Scope               string
		Version             string
		HistogramBoundaries []float64 `json:"histogram_boundaries"`
		Metrics             map[string]string
		Operations          []struct {
			Name, Destination, Error, Type string
			Transaction                    bool
			Count                          uint64
		} `json:"scenario_operations"`
		Received struct{ First, Redelivered int64 }
		Sent     []struct {
			Destination, Error string
			Transaction        bool
			Count              int64
		}
	}
	require.NoError(t, json.Unmarshal(raw, &contract))
	require.Len(t, data.ScopeMetrics, 1)
	scope := data.ScopeMetrics[0]
	require.Equal(t, contract.Scope, scope.Scope.Name)
	require.Equal(t, contract.Version, scope.Scope.Version)
	var operations []metricdata.HistogramDataPoint[float64]
	var sent []metricdata.DataPoint[int64]
	totals := map[bool]int64{}
	for _, m := range scope.Metrics {
		require.Equal(t, contract.Metrics[m.Name], m.Unit)
		switch m.Name {
		case "messaging.client.operation.duration":
			operations = m.Data.(metricdata.Histogram[float64]).DataPoints
		case "messaging.client.sent.messages":
			sent = m.Data.(metricdata.Sum[int64]).DataPoints
		case "messaging.client.consumed.messages":
			for _, p := range m.Data.(metricdata.Sum[int64]).DataPoints {
				v, _ := p.Attributes.Value(attribute.Key("postgremq.redelivered"))
				totals[v.AsBool()] += p.Value
			}
		}
	}
	require.Len(t, sent, len(contract.Sent))
	for _, expected := range contract.Sent {
		found := false
		for _, point := range sent {
			attrs := []attribute.KeyValue{
				attribute.String("messaging.system", "postgremq"), attribute.String("messaging.operation.name", "publish"),
				attribute.String("messaging.operation.type", "send"), attribute.String("messaging.destination.name", expected.Destination),
				attribute.Bool("postgremq.transaction", expected.Transaction),
			}
			if expected.Error != "" {
				attrs = append(attrs, attribute.String("error.type", expected.Error))
			}
			set := attribute.NewSet(attrs...)
			if point.Attributes.Equals(&set) {
				found = true
				require.Equal(t, expected.Count, point.Value)
			}
		}
		require.True(t, found, "missing sent counter: %+v", expected)
	}
	require.Len(t, operations, len(contract.Operations))
	for _, expected := range contract.Operations {
		found := false
		for _, p := range operations {
			get := func(k string) attribute.Value { v, _ := p.Attributes.Value(attribute.Key(k)); return v }
			if get("messaging.operation.name").AsString() != expected.Name || get("messaging.destination.name").AsString() != expected.Destination || get("postgremq.transaction").AsBool() != expected.Transaction || get("error.type").AsString() != expected.Error {
				continue
			}
			found = true
			require.Equal(t, expected.Count, p.Count)
			require.GreaterOrEqual(t, p.Sum, 0.0)
			require.Equal(t, contract.HistogramBoundaries, p.Bounds)
			require.Equal(t, "postgremq", get("messaging.system").AsString())
			require.Equal(t, expected.Type, get("messaging.operation.type").AsString())
			require.Equal(t, 5+boolInt(expected.Error != ""), p.Attributes.Len(), "bounded attributes only")
		}
		require.True(t, found, "missing metric: %+v", expected)
	}
	require.Equal(t, contract.Received.First, totals[false])
	require.Equal(t, contract.Received.Redelivered, totals[true])
}

func boolInt(v bool) int {
	if v {
		return 1
	}
	return 0
}

func TestMetricsCountLogicalOperationAcrossRetry(t *testing.T) {
	pool, ctx := setupTestConnection(t)
	reader := sdkmetric.NewManualReader()
	provider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(reader))
	defer func() { require.NoError(t, provider.Shutdown(ctx)) }()
	conn, err := postgremq.DialFromPool(pool, postgremq.WithMeterProvider(provider))
	require.NoError(t, err)
	defer func() { require.NoError(t, conn.Close()) }()
	require.NoError(t, conn.CreateTopic(ctx, "metrics"))
	sql, err := os.ReadFile("../observability/retry-once.sql")
	require.NoError(t, err)
	_, err = pool.Exec(ctx, string(sql))
	require.NoError(t, err)
	_, err = conn.Publish(ctx, "metrics", []byte(`{}`))
	require.NoError(t, err)
	var attempts int
	require.NoError(t, pool.QueryRow(ctx, "SELECT last_value FROM public.metrics_retry_probe").Scan(&attempts))
	require.Equal(t, 2, attempts)
	var data metricdata.ResourceMetrics
	require.NoError(t, reader.Collect(ctx, &data))
	var points []metricdata.HistogramDataPoint[float64]
	attemptsByError := map[string]int64{}
	for _, m := range data.ScopeMetrics[0].Metrics {
		switch m.Name {
		case "messaging.client.operation.duration":
			points = m.Data.(metricdata.Histogram[float64]).DataPoints
		case "messaging.client.sent.messages":
			for _, point := range m.Data.(metricdata.Sum[int64]).DataPoints {
				code, _ := point.Attributes.Value(attribute.Key("error.type"))
				attemptsByError[code.AsString()] += point.Value
			}
		}
	}
	require.Equal(t, map[string]int64{"": 1, "other": 1}, attemptsByError)

	require.Len(t, points, 1)
	require.EqualValues(t, 1, points[0].Count)
	_, hasError := points[0].Attributes.Value(attribute.Key("error.type"))
	require.False(t, hasError)
	require.GreaterOrEqual(t, points[0].Sum, 0.1, "duration includes retry backoff")
}

// A global SDK does not implicitly enable telemetry; connection shutdown never
// owns the application provider, whether explicitly supplied or omitted.
func TestMetricsProviderIsOptionalAndApplicationOwned(t *testing.T) {
	for _, enabled := range []bool{false, true} {
		t.Run(map[bool]string{false: "omitted", true: "provided"}[enabled], func(t *testing.T) {
			pool, ctx := setupTestConnection(t)
			reader := sdkmetric.NewManualReader()
			provider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(reader))
			defer func() { require.NoError(t, provider.Shutdown(ctx)) }()
			previous := otel.GetMeterProvider()
			otel.SetMeterProvider(provider)
			defer otel.SetMeterProvider(previous)
			var opts []postgremq.ConnectionOption
			if enabled {
				opts = append(opts, postgremq.WithMeterProvider(provider))
			}
			conn, err := postgremq.DialFromPool(pool, opts...)
			require.NoError(t, err)
			defer func() { require.NoError(t, conn.Close()) }()
			require.NoError(t, conn.CreateTopic(ctx, "optional"))
			require.NoError(t, conn.CreateQueue(ctx, "q", "optional", false))
			_, err = conn.Publish(ctx, "optional", []byte(`{}`))
			require.NoError(t, err)
			messages, err := conn.ConsumeMessages(ctx, "q", 1, 30)
			require.NoError(t, err)
			require.Len(t, messages, 1)
			require.NoError(t, messages[0].Ack(ctx))
			require.NoError(t, conn.Close())
			// This API rejection must not increment the attempted-send counter.
			_, err = conn.Publish(ctx, "optional", []byte(`{}`))
			require.ErrorIs(t, err, postgremq.ErrConnectionClosed)
			probe, err := provider.Meter("application").Int64Counter("application.probe")
			require.NoError(t, err)
			probe.Add(ctx, 1)
			var data metricdata.ResourceMetrics
			require.NoError(t, reader.Collect(ctx, &data))
			foundClient := false
			foundProbe := false
			for _, scope := range data.ScopeMetrics {
				if scope.Scope.Name == "application" {
					foundProbe = true
					require.EqualValues(t, 1, scope.Metrics[0].Data.(metricdata.Sum[int64]).DataPoints[0].Value)
				}
				if scope.Scope.Name == "postgremq" {
					foundClient = true
					for _, m := range scope.Metrics {
						if m.Name == "messaging.client.sent.messages" {
							points := m.Data.(metricdata.Sum[int64]).DataPoints
							require.Len(t, points, 1)
							require.EqualValues(t, 1, points[0].Value)
						}
					}
				}
			}
			require.True(t, foundProbe)
			require.Equal(t, enabled, foundClient)
		})
	}
}
